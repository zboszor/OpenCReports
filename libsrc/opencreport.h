/*
 * OpenCReports main header
 * Copyright (C) 2019 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */
#ifndef _OPENCREPORT_H_
#define _OPENCREPORT_H_

#include <stdbool.h>
#include <stdint.h>
#include <time.h>
#include <mpfr.h>

/* Main report structure type */
struct opencreport;
typedef struct opencreport opencreport;

struct ocrpt_expr;
typedef struct ocrpt_expr ocrpt_expr;

enum ocrpt_result_type {
	/*
	 * Error in parsing or evaluation
	 */
	OCRPT_RESULT_ERROR,

	/*
	 * Constants
	 */
	OCRPT_RESULT_STRING,
	OCRPT_RESULT_NUMBER,
	OCRPT_RESULT_DATETIME
};

struct ocrpt_string {
	char *str;
	size_t allocated_len;
	size_t len;
};
typedef struct ocrpt_string ocrpt_string;

struct ocrpt_result {
	enum ocrpt_result_type type;
	/* Original lexer token or (computed) string value for expression */
	ocrpt_string *string;
	/* Converted numeric constant or computed numeric value for expression */
	mpfr_t number;
	bool number_initialized;
	bool string_owned;
	bool isnull;
	/* Datetime value */
	struct tm datetime;
};
typedef struct ocrpt_result ocrpt_result;

struct ocrpt_query;
typedef struct ocrpt_query ocrpt_query;

struct ocrpt_query_result {
	const char *name;
	bool name_allocated;
	ocrpt_result result;
};
typedef struct ocrpt_query_result ocrpt_query_result;

struct ocrpt_input {
	void (*describe)(ocrpt_query *, ocrpt_query_result **, int32_t *);
	void (*rewind)(ocrpt_query *);
	bool (*next)(ocrpt_query *);
	bool (*populate_result)(ocrpt_query *);
	bool (*isdone)(ocrpt_query *);
	void (*free)(ocrpt_query *);
};

typedef struct ocrpt_input ocrpt_input;

struct ocrpt_datasource;
typedef struct ocrpt_datasource ocrpt_datasource;

struct ocrpt_paper {
	const char *name;
	double width;
	double height;
};
typedef struct ocrpt_paper ocrpt_paper;

#if 0
struct report_line;
typedef struct report_line report_line;

struct ocrpt_report_part;
typedef struct ocrpt_report_part ocrpt_report_part;
#endif

#define OCRPT_MPFR_PRECISION_BITS	(256)

/*
 * Create a new empty OpenCReports structure
 */
opencreport *ocrpt_init(void);
/*
 * Free an OpenCReports structure
 */
void ocrpt_free(opencreport *o);
/*
 * Set MPFR numeric precision
 */
void ocrpt_set_numeric_precision_bits(opencreport *o, mpfr_prec_t prec);
/*
 * Set MPFR rounding mode
 */
void ocrpt_set_rounding_mode(opencreport *o, mpfr_rnd_t rndmode);

/********************************
 * Expression related functions *
 ********************************/

/*
 * Create an expression parse tree from an expression string
 */
ocrpt_expr *ocrpt_expr_parse(opencreport *o, const char *str, char **err);
/*
 * Optimize expression after parsing
 */
void ocrpt_expr_optimize(opencreport *o, ocrpt_expr *e);
/*
 * Resolve variable references in the expression
 */
void ocrpt_expr_resolve(opencreport *o, ocrpt_expr *e);
/*
 * Evaluate the expression, i.e. compute its ocrpt_result
 * It must be called after ocrpt_navigate_next(), see below.
 */
ocrpt_result *ocrpt_expr_eval(opencreport *o, ocrpt_expr *e);
/*
 * Print an expression on stdout. Good for unit testing.
 */
void ocrpt_expr_print(opencreport *o, ocrpt_expr *e);
/*
 * Print the result data. Good for unit testing.
 */
void ocrpt_expr_result_print(ocrpt_result *r);
/*
 * Count the number of expression nodes
 */
int ocrpt_expr_nodes(ocrpt_expr *e);
/*
 * Free an expression parse tree
 */
void ocrpt_free_expr(ocrpt_expr *e);

/****************************
 * Memory handling wrappers *
 ****************************/

typedef void *(*ocrpt_mem_malloc_t)(size_t);
typedef void *(*ocrpt_mem_realloc_t)(void *, size_t);
typedef void (*ocrpt_mem_free_t)(const void *);
typedef char *(*ocrpt_mem_strdup_t)(const char *);
typedef char *(*ocrpt_mem_strndup_t)(const char *, size_t);

extern ocrpt_mem_malloc_t ocrpt_mem_malloc0;
extern ocrpt_mem_realloc_t ocrpt_mem_realloc0;
extern ocrpt_mem_free_t ocrpt_mem_free0;
extern ocrpt_mem_strdup_t ocrpt_mem_strdup0;
extern ocrpt_mem_strndup_t ocrpt_mem_strndup0;

static inline void *ocrpt_mem_malloc(size_t sz) __attribute__((malloc,alloc_size(1)));
static inline void *ocrpt_mem_malloc(size_t sz) { return ocrpt_mem_malloc0(sz); }

static inline void *ocrpt_mem_realloc(void *ptr, size_t sz) __attribute__((alloc_size(2)));
static inline void *ocrpt_mem_realloc(void *ptr, size_t sz) { return ocrpt_mem_realloc0(ptr, sz); }

static inline void ocrpt_mem_free(const void *ptr) { ocrpt_mem_free0(ptr); }
static inline void ocrpt_strfree(const char *s) { ocrpt_mem_free0((const void *)s); }

static inline void *ocrpt_mem_strdup(const char *ptr) { return (ptr ? ocrpt_mem_strdup0(ptr) : NULL); }

static inline void *ocrpt_mem_strndup(const char *ptr, size_t sz) { return (ptr ? ocrpt_mem_strndup0(ptr, sz) : NULL); }

void ocrpt_mem_set_alloc_funcs(ocrpt_mem_malloc_t rmalloc,
								ocrpt_mem_realloc_t rrealloc,
								ocrpt_mem_free_t rfree,
								ocrpt_mem_strdup_t rstrdup,
								ocrpt_mem_strndup_t rstrndup);

/*
 * String related functions
 */
ocrpt_string *ocrpt_mem_string_new(const char *str, bool copy);
ocrpt_string *ocrpt_mem_string_new_with_len(const char *str, size_t len);
char *ocrpt_mem_string_free(ocrpt_string *string, bool free_str);
void ocrpt_mem_string_append_len(ocrpt_string *string, const char *str, const size_t len);
void ocrpt_mem_string_append(ocrpt_string *string, const char *str);
void ocrpt_mem_string_append_c(ocrpt_string *string, const char c);

/********************************
 * Paper size related functions *
 ********************************/

/*
 * System default paper name and size
 */
const ocrpt_paper *ocrpt_get_system_paper(void);
/*
 * Get paper size of the specified paper name
 */
const ocrpt_paper *ocrpt_get_paper_by_name(const char *paper);
/*
 * Set paper for the report using an ocrpt_paper structure
 * The contents of the structure is copied.
 */
void ocrpt_set_paper(opencreport *o, const ocrpt_paper *paper);
/*
 * Set paper for the report using a paper name
 * If the paper name is unknown, the system default paper is set.
 */
void ocrpt_set_paper_by_name(opencreport *o, const char *paper);
/*
 * Get the report's currently set paper
 */
const ocrpt_paper *ocrpt_get_paper(opencreport *o);
/*
 * Iterator for supported paper names and sizes
 * The returned paper structures are library-global
 * but the iterator is per report.
 */
const ocrpt_paper *ocrpt_paper_first(opencreport *o);
const ocrpt_paper *ocrpt_paper_next(opencreport *o);

/******************************************
 * Datasource and query related functions *
 ******************************************/

/*
 * Add a custom datasource not covered by
 * the currently implemented sources.
 */
ocrpt_datasource *ocrpt_add_datasource(opencreport *o, const char *source_name, const ocrpt_input *input);
/*
 * Find the datasource using its name
 */
ocrpt_datasource *ocrpt_find_datasource(opencreport *o, const char *source_name);
/*
 * Validate the datasource pointer against the report structure
 */
ocrpt_datasource *ocrpt_validate_datasource(opencreport *o, ocrpt_datasource *source);
/*
 * Add an array datasource
 *
 * Calling this is optional, as an array datasource called
 * "array" is automatically added to an opencreport structure.
 */
ocrpt_datasource *ocrpt_add_array_datasource(opencreport *o, const char *source_name);
/*
 * Add an array query using the datasource pointer
 */
ocrpt_query *ocrpt_add_array_query(opencreport *o, ocrpt_datasource *source,
									const char *name, void *array,
									int32_t rows, int32_t cols,
									const enum ocrpt_result_type *types);
/*
 * Add an array query using the datasource name
 */
ocrpt_query *ocrpt_add_array_query_as(opencreport *o, const char *source_name,
									const char *name, void *array,
									int32_t rows, int32_t cols,
									const enum ocrpt_result_type *types);
/*
 * Return the query result array and the number of columns in it
 *
 * It must be re-run for every new data source row since
 * the pointer is invalidated after ocrpt_navigate_next()
 */
ocrpt_query_result *ocrpt_query_get_result(ocrpt_query *q, int32_t *cols);
/*
 * Add follower query with a match function
 */
bool ocrpt_add_query_follower_n_to_1(opencreport *o,
									ocrpt_query *leader,
									ocrpt_query *follower,
									ocrpt_expr *match);
/*
 * Add follower query (runs side-by-side with "leader")
 */
bool ocrpt_add_query_follower(opencreport *o,
									ocrpt_query *leader,
									ocrpt_query *follower);
/*
 * Free a query and remove it from follower references
 */
void ocrpt_free_query(opencreport *o, ocrpt_query *q);
/*
 * Start query navigation from the beginning of the resultset
 */
void ocrpt_navigate_start(opencreport *o, ocrpt_query *q);
/*
 * Move to next row in the query resultset
 */
bool ocrpt_navigate_next(opencreport *o, ocrpt_query *q);

#if 0
/*
 * Create a report part
 */
ocrpt_report_part *ocrpt_create_report_part(void);
/*
 * Add report header to report part
 */
ocrpt_report_part_set_report_header(ocrpt_report_part *part, ...);
#endif
/*
 * Add XML report description from a file
 */
int ocrpt_add_report(opencreport *o, const char *filename);
/*
 * Add XML report description using a memory buffer (zero-terminated string)
 */
int ocrpt_add_report_from_buffer(opencreport *o, const char *buffer);
/*
 * Parse the reports added up to this point.
 * Allow no deviation from the OpenCReports DTD.
 */
int ocrpt_parse(opencreport *o);
/*
 * Parse the reports added up to this point.
 * Optionally allow deviation from the DTD.
 */
int ocrpt_parse2(opencreport *o, int allow_bad_xml);
/*
 * Execute the reports added up to this point.
 * Parse any unparsed report XMLs with non-strict checks.
 */
int ocrpt_execute(opencreport *o);

#endif /* _OPENCREPORT_H_ */
