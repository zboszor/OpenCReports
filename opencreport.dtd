<!ELEMENT OpenCReport (Datasources?,Queries?,(Part|Report)*)>
<!ELEMENT Datasources (Datasource+)>
<!ELEMENT Datasource EMPTY>
<!ATTLIST Datasource
	name CDATA #REQUIRED
	type CDATA #REQUIRED
	host CDATA #IMPLIED
	unix_socket CDATA #IMPLIED
	port CDATA #IMPLIED
	dbname CDATA #IMPLIED
	user CDATA #IMPLIED
	password CDATA #IMPLIED
	connstr CDATA #IMPLIED
	optionfile CDATA #IMPLIED
	group CDATA #IMPLIED >
<!ELEMENT Queries (Query)+>
<!ELEMENT Query (#PCDATA)>
<!ATTLIST Query
	name CDATA #REQUIRED
	datasource CDATA #REQUIRED
	value CDATA #IMPLIED
	rows CDATA #IMPLIED
	cols CDATA #IMPLIED
	coltypes CDATA #IMPLIED
	follower_for CDATA #IMPLIED
	follower_expr CDATA #IMPLIED >
<!ELEMENT Part (PageHeader|PageFooter|ReportHeader|ReportFooter|pr)*>
<!ATTLIST Part
	layout CDATA #IMPLIED
	pagesAcross CDATA #IMPLIED
	pages_across CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	font_size CDATA #IMPLIED
	orientation CDATA #IMPLIED
	topMargin CDATA #IMPLIED
	top_margin CDATA #IMPLIED
	leftMargin CDATA #IMPLIED
	left_margin CDATA #IMPLIED
	bottomMargin CDATA #IMPLIED
	bottom_margin CDATA #IMPLIED
	paperType CDATA #IMPLIED
	paper_type CDATA #IMPLIED
	iterations CDATA #IMPLIED
	suppressPageHeaderFirstPage CDATA #IMPLIED >
<!ELEMENT pr (pd)+>
<!ATTLIST pr
	layout CDATA #IMPLIED
	newpage CDATA #IMPLIED >
<!ELEMENT pd (Report*)>
<!ATTLIST pd
	width CDATA #IMPLIED
	height CDATA #IMPLIED
	border_width CDATA #IMPLIED
	border_color CDATA #IMPLIED >
<!ELEMENT Report ((Alternate|NoData)|(Breaks?,Variables?)|PageHeader|PageFooter|ReportHeader|ReportFooter|(Detail|Graph|Chart)*|xi:include)* >
<!ATTLIST Report
	pagesAcross CDATA #IMPLIED
	pages_across CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	font_size CDATA #IMPLIED
	orientation CDATA #IMPLIED
	topMargin CDATA #IMPLIED
	top_margin CDATA #IMPLIED
	leftMargin CDATA #IMPLIED
	left_margin CDATA #IMPLIED
	bottomMargin CDATA #IMPLIED
	bottom_margin CDATA #IMPLIED
	paperType CDATA #IMPLIED
	paper_type CDATA #IMPLIED
	iterations CDATA #IMPLIED
	suppressPageHeaderFirstPage CDATA #IMPLIED

	query CDATA #IMPLIED
	xmlns:xi CDATA #IMPLIED
	suppress CDATA #IMPLIED
	uniquerow CDATA #IMPLIED
	detail_columns CDATA #IMPLIED
	column_pad CDATA #IMPLIED >
<!ELEMENT xi:include EMPTY>
<!ATTLIST xi:include
	href CDATA #REQUIRED >
<!ELEMENT Alternate (NoData) >
<!ELEMENT NoData (Output) >
<!ELEMENT ReportHeader (Output)* >
<!ELEMENT PageHeader (Output)* >
<!ELEMENT ReportFooter (Output)* >
<!ELEMENT PageFooter (Output)* >
<!ELEMENT Output (Line | HorizontalLine | Image)* >
<!ELEMENT Line (field | literal)* >
<!ATTLIST Line
	fontSize CDATA #IMPLIED
	bold CDATA #IMPLIED
	suppress CDATA #IMPLIED
	color CDATA #IMPLIED
	bgcolor CDATA #IMPLIED >
<!ELEMENT HorizontalLine EMPTY>
<!ATTLIST HorizontalLine
	size CDATA #IMPLIED
	indent CDATA #IMPLIED
	length CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	suppress CDATA #IMPLIED
	bgcolor CDATA #IMPLIED >
<!ELEMENT Image EMPTY >
<!ATTLIST Image
	value CDATA #REQUIRED
	type CDATA #REQUIRED
	width CDATA #REQUIRED
	height CDATA #REQUIRED >
<!ELEMENT field EMPTY>
<!ATTLIST field
	value CDATA #REQUIRED
	delayed CDATA #IMPLIED
	precalculate CDATA #IMPLIED
	width CDATA #IMPLIED
	align CDATA #IMPLIED
	format CDATA #IMPLIED
	bgcolor CDATA #IMPLIED
	color CDATA #IMPLIED
	link CDATA #IMPLIED
	maxlines CDATA #IMPLIED
	col CDATA #IMPLIED >
<!ELEMENT literal (#PCDATA) >
<!ATTLIST literal
	align CDATA #IMPLIED
	bgcolor CDATA #IMPLIED
	color CDATA #IMPLIED
	width CDATA #IMPLIED
	bold CDATA #IMPLIED
	italics CDATA #IMPLIED
	link CDATA #IMPLIED
	translate CDATA #IMPLIED
	col CDATA #IMPLIED >
<!ELEMENT Detail (FieldHeaders?, FieldDetails?) >
<!ELEMENT FieldHeaders (Output)* >
<!ELEMENT FieldDetails (Output)* >
<!ELEMENT Graph (Plot+) >
<!ATTLIST Graph
	name CDATA #IMPLIED
	type CDATA #REQUIRED
	subtype CDATA #REQUIRED
	width CDATA #IMPLIED
	height CDATA #IMPLIED
	bold_titles CDATA #IMPLIED
	title CDATA #IMPLIED
	legend_bg_color CDATA #IMPLIED
	legend_orientation CDATA #IMPLIED
	draw_x_line CDATA #IMPLIED
	draw_y_line CDATA #IMPLIED
	grid_color CDATA #IMPLIED
	x_axis_title CDATA #IMPLIED
	y_axis_title CDATA #IMPLIED
	y_axis_mod CDATA #IMPLIED
	y_axis_title_right CDATA #IMPLIED
	y_axis_decimals CDATA #IMPLIED
	y_axis_decimals_right CDATA #IMPLIED >
<!ELEMENT Plot EMPTY>
<!ATTLIST Plot
	axis CDATA #IMPLIED
	field CDATA #IMPLIED
	label CDATA #IMPLIED
	side CDATA #IMPLIED
	disabled CDATA #IMPLIED
	color CDATA #IMPLIED >
<!ELEMENT Breaks (Break)* >
<!ELEMENT Break (BreakHeader | BreakFooter | BreakFields)* >
<!ATTLIST Break
	name CDATA #REQUIRED
	headernewpage CDATA #IMPLIED
	suppressblank CDATA #IMPLIED
	newpage CDATA #IMPLIED >
<!ELEMENT BreakHeader (Output)* >
<!ELEMENT BreakFooter (Output)* >
<!ELEMENT BreakFields (BreakField)* >
<!ELEMENT BreakField EMPTY>
<!ATTLIST BreakField
	value CDATA #REQUIRED >
<!ELEMENT Variables (Variable)* >
<!ELEMENT Variable EMPTY >
<!ATTLIST Variable
	name CDATA #REQUIRED
	value CDATA #REQUIRED
	type (expression|count|sum|average|lowest|highest) "sum"
	resetonbreak CDATA #IMPLIED
	precalculate CDATA #IMPLIED
	delayed CDATA #IMPLIED >
