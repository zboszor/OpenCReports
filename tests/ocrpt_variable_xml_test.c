/*
 * OpenCReports test
 * Copyright (C) 2021-2021 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

#include <stdio.h>

#include <opencreport.h>
#include "ocrpt_test_common.c"

/* NOT static */ const char *array[4][5] = {
	{ "id", "name", "property", "age", "adult" },
	{ "1", "Fred Flintstone", "strong", "31", "yes" },
	{ "2", "Wilma Flintstone", "charming", "28", "yes" },
	{ "3", "Pebbles Flintstone", "young", "5e-1", "no" }
};

/* NOT static */ const enum ocrpt_result_type coltypes[5] = {
	OCRPT_RESULT_NUMBER, OCRPT_RESULT_STRING, OCRPT_RESULT_STRING, OCRPT_RESULT_NUMBER, OCRPT_RESULT_NUMBER
};

struct test_expr_types {
	const char *type_name;
	ocrpt_var_type type;
};

static char *test_vars[] = {
	"v.var_id_count",
	"v.var_id_countall",
	"v.var_id_sum",
	"v.var_id_highest",
	"v.var_id_lowest",
	"v.var_age_avg"
};

#define N_TEST_VARS (sizeof(test_vars) / sizeof(char *))

int main(void) {
	opencreport *o = ocrpt_init();
	ocrpt_query *q;
	ocrpt_query_result *qr;
	ocrpt_report *r;
	ocrpt_expr *e[N_TEST_VARS];
	int32_t cols, row, i;

	if (!ocrpt_parse_xml(o, "ocrpt_variable_xml_test.xml")) {
		printf("XML parse error\n");
		ocrpt_free(o);
		return 0;
	}

	q = ocrpt_query_find(o, "a");
	qr = ocrpt_query_get_result(q, &cols);

	for (i = 0; i < N_TEST_VARS; i++)
		e[i] = ocrpt_expr_parse(o, test_vars[i], NULL);

	/* There is only one ocrpt_report pointer in o->parts, extract it. */
	r = (ocrpt_report *)((ocrpt_list *)(((ocrpt_part *)(o->parts->data))->rows->data))->data;

	row = 0;
	ocrpt_query_navigate_start(o, q);
	ocrpt_report_resolve_variables(o, r);

	for (i = 0; i < N_TEST_VARS; i++) {
		ocrpt_expr_resolve(o, r, e[i]);
		ocrpt_expr_optimize(o, r, e[i]);
	}

	while (ocrpt_query_navigate_next(o, q)) {
		ocrpt_result *rs;

		qr = ocrpt_query_get_result(q, &cols);

		ocrpt_report_evaluate_variables(o, r);

		printf("Row #%d\n", row++);
		print_result_row("a", qr, cols);

		printf("\n");

		ocrpt_report_evaluate_variables(o, r);

		for (i = 0; i < N_TEST_VARS; i++) {
			printf("Expression: ");
			ocrpt_expr_print(o, e[i]);
			rs = ocrpt_expr_eval(o, r, e[i]);
			printf("Evaluated: ");
			ocrpt_result_print(rs);
		}

		printf("\n");
	}

	for (i = 0; i < N_TEST_VARS; i++)
		ocrpt_expr_free(e[i]);

	ocrpt_free(o);

	return 0;
}
