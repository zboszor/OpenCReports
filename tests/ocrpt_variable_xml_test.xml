<?xml version="1.0"?>
<!DOCTYPE OpenCReport SYSTEM "opencreport.dtd">
<OpenCReport>
	<Datasources>
		<Datasource name="myarray" type="'array'" />
	</Datasources>
	<Queries>
		<Query name="a" datasource="myarray" value="'array'" coltypes="'coltypes'" rows="3" cols="6" />
	</Queries>
	<Report>
		<Variables>
			<Variable name="var_id_count" value="id" type="count" />
			<Variable name="var_id_countall" value="id" type="countall" />
			<Variable name="var_id_sum" value="id" type="sum" />
			<Variable name="var_id_highest" value="id" type="highest" />
			<Variable name="var_id_lowest" value="id" type="lowest" />
			<Variable name="var_id_avg" value="id" type="average" />
			<Variable name="var_id_avgall" value="id" type="averageall" />
			<Variable name="var_id2_count" value="id2" type="count" />
			<Variable name="var_id2_countall" value="id2" type="countall" />
			<Variable name="var_id2_sum" value="id2" type="sum" />
			<Variable name="var_id2_highest" value="id2" type="highest" />
			<Variable name="var_id2_lowest" value="id2" type="lowest" />
			<Variable name="var_id2_avg" value="id2" type="average" />
			<Variable name="var_id2_avgall" value="id2" type="averageall" />
		</Variables>
	</Report>
</OpenCReport>
