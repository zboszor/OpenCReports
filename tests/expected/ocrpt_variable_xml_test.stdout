Row #0
Query: 'a':
	Col #0: 'id': string value: NULL (converted to number: 1.000000)
	Col #1: 'id2': string value: NULL (converted to number: 1.000000)
	Col #2: 'name': string value: Fred Flintstone
	Col #3: 'property': string value: strong
	Col #4: 'age': string value: NULL (converted to number: 31.000000)
	Col #5: 'adult': string value: NULL (converted to number: 1.000000)

Expression: .'id2'
Evaluated: (number)1.000000
Expression: v.'var_id_count'
Evaluated: (number)1.000000
Expression: v.'var_id_countall'
Evaluated: (number)1.000000
Expression: v.'var_id_sum'
Evaluated: (number)1.000000
Expression: v.'var_id_highest'
Evaluated: (number)1.000000
Expression: v.'var_id_lowest'
Evaluated: (number)1.000000
Expression: v.'var_id_avg'
Evaluated: (number)1.000000
Expression: v.'var_id_avgall'
Evaluated: (number)1.000000
Expression: v.'var_id2_count'
Evaluated: (number)1.000000
Expression: v.'var_id2_countall'
Evaluated: (number)1.000000
Expression: v.'var_id2_sum'
Evaluated: (number)1.000000
Expression: v.'var_id2_highest'
Evaluated: (number)1.000000
Expression: v.'var_id2_lowest'
Evaluated: (number)1.000000
Expression: v.'var_id2_avg'
Evaluated: (number)1.000000
Expression: v.'var_id2_avgall'
Evaluated: (number)1.000000

Row #1
Query: 'a':
	Col #0: 'id': string value: NULL (converted to number: 2.000000)
	Col #1: 'id2': string value: NULL
	Col #2: 'name': string value: Wilma Flintstone
	Col #3: 'property': string value: charming
	Col #4: 'age': string value: NULL (converted to number: 28.000000)
	Col #5: 'adult': string value: NULL (converted to number: 1.000000)

Expression: .'id2'
Evaluated: (number)NULL
Expression: v.'var_id_count'
Evaluated: (number)2.000000
Expression: v.'var_id_countall'
Evaluated: (number)2.000000
Expression: v.'var_id_sum'
Evaluated: (number)3.000000
Expression: v.'var_id_highest'
Evaluated: (number)2.000000
Expression: v.'var_id_lowest'
Evaluated: (number)1.000000
Expression: v.'var_id_avg'
Evaluated: (number)1.500000
Expression: v.'var_id_avgall'
Evaluated: (number)1.500000
Expression: v.'var_id2_count'
Evaluated: (number)1.000000
Expression: v.'var_id2_countall'
Evaluated: (number)2.000000
Expression: v.'var_id2_sum'
Evaluated: (number)1.000000
Expression: v.'var_id2_highest'
Evaluated: (number)1.000000
Expression: v.'var_id2_lowest'
Evaluated: (number)1.000000
Expression: v.'var_id2_avg'
Evaluated: (number)1.000000
Expression: v.'var_id2_avgall'
Evaluated: (number)0.500000

Row #2
Query: 'a':
	Col #0: 'id': string value: NULL (converted to number: 3.000000)
	Col #1: 'id2': string value: NULL (converted to number: 3.000000)
	Col #2: 'name': string value: Pebbles Flintstone
	Col #3: 'property': string value: young
	Col #4: 'age': string value: NULL (converted to number: 0.500000)
	Col #5: 'adult': string value: NULL (converted to number: 0.000000)

Expression: .'id2'
Evaluated: (number)3.000000
Expression: v.'var_id_count'
Evaluated: (number)3.000000
Expression: v.'var_id_countall'
Evaluated: (number)3.000000
Expression: v.'var_id_sum'
Evaluated: (number)6.000000
Expression: v.'var_id_highest'
Evaluated: (number)3.000000
Expression: v.'var_id_lowest'
Evaluated: (number)1.000000
Expression: v.'var_id_avg'
Evaluated: (number)2.000000
Expression: v.'var_id_avgall'
Evaluated: (number)2.000000
Expression: v.'var_id2_count'
Evaluated: (number)2.000000
Expression: v.'var_id2_countall'
Evaluated: (number)3.000000
Expression: v.'var_id2_sum'
Evaluated: (number)4.000000
Expression: v.'var_id2_highest'
Evaluated: (number)3.000000
Expression: v.'var_id2_lowest'
Evaluated: (number)1.000000
Expression: v.'var_id2_avg'
Evaluated: (number)2.000000
Expression: v.'var_id2_avgall'
Evaluated: (number)1.333333

